### PR TITLE
Add System.Diagnostics.TextMapPropagator support

### LINKED ISSUE #27237: Allow control over Activity creation
### Is your feature request related to a problem? Please describe.

It is currently not possible to override (or suppress) creation of a new `System.Diagnostics.Activity` since it happens as a part of hosting.

### Describe the solution you'd like

We should move Activity creation from hosting into a new middleware

### Additional context

Customer quote:
> Our service is in the middle of a bunch of other services which all use a Http header, `x-ms-correlation-id` with the value being a Guid formatted with dashes, to pass correlation for aligning traces and logs. I want to just reuse their Id without my developers having to even care, but if I let ASP.NET do its magic, there is no incoming traceparent  header so a new activity is created from scratch (not calling SetParentId beforehand).  This means and I have to do a bunch of gymnastics to grab `x-ms-correlation-id` and attach it as baggage and ensure I log it as well with everything (so now I have 2 “traceids”).  I just want to be able to reuse an incoming traceid/guid and let Activity do all the hard work propagating and adding to logs.

### LINKED ISSUE #30392: Ability to disable creation of intermediate activities in DiagnosticsHandler
### The Problem

AspNetCore services are creating intermediate activities during HTTP comms, when invoking the service from any external component that has telemetry instrumentation the traces end up as broken hierarchies as unintended child spans are being created in the aspnetcore  service, especially if said service sits as a bridge between 2 components.

for example, given that services **A**, **C** and **D** are instrumented but not **B**

A -> B - parent **abc** span **def**
B -> C - parent **def** span **ghi**
C -> D - parent **ghi** span **jkl**

if I have no intention of instrumenting B then preferably it won't create a span, otherwise the exported trace ends up broken and instead of looking like this

**abc** / **def** / **ghi** / **jkl** /

it will look like this

**abc** / **def** / 
**jkl** /

additionally, this creates what can be deemed as unnecessary noise in the tracing and is potentially a performance consideration for highly optimised high throughput systems (questionable).

### Suggested Solution
I'm not sure of the implications this would have on users relying on this as a default behaviour so perhaps an opt-in may not be viable but at least the ability to opt out.

### Additional Context
Personally I experienced this issue when using Dapr sidecars. For the most part, the sidecars themselves emit the desired telemetry however the traces all end up broken when invoking aspnetcore services because of the elusive spans.

**Intermediate activities/spans not exported**
image

**Intermediate activities/spans exported**
image

### PR DESCRIPTION
**Pull request by shirhatti**

### CODE REVIEW

**Comment by shirhatti** [comment]
plaintext on `asp-citrine-lin`

| application           | old                        | new                        |         |
| --------------------- | -------------------------- | -------------------------- | ------- |
| CPU Usage (%)         |                         97 |                         99 |  +2.06% |
| Cores usage (%)       |                      2,730 |                      2,761 |  +1.14% |
| Working Set (MB)      |                        374 |                        365 |  -2.41% |
| Private Memory (MB)   |                        949 |                        939 |  -1.05% |
| Build Time (ms)       |                      2,883 |                      2,897 |  +0.49% |
| Start Time (ms)       |                        141 |                        168 | +19.15% |
| Published Size (KB)   |                    106,148 |                    106,148 |   0.00% |
| .NET Core SDK Version | 6.0.100-preview.6.21276.12 | 6.0.100-preview.6.21276.12 |         |


| load                   | old        | new        |         |
| ---------------------- | ---------- | ---------- | ------- |
| CPU Usage (%)          |         31 |         31 |   0.00% |
| Cores usage (%)        |        863 |        880 |  +1.97% |
| Working Set (MB)       |         38 |         37 |  -2.63% |
| Private Memory (MB)    |        358 |        358 |   0.00% |
| Start Time (ms)        |          0 |          0 |         |
| First Request (ms)     |         69 |         69 |   0.00% |
| Requests/sec           |  3,898,513 |  4,019,992 |  +3.12% |
| Requests               | 58,863,739 | 60,700,594 |  +3.12% |
| Mean latency (ms)      |       0.83 |       0.79 |  -4.09% |
| Max latency (ms)       |      55.76 |      48.20 | -13.56% |
| Bad responses          |          0 |          0 |         |
| Socket errors          |          0 |          0 |         |
| Read throughput (MB/s) |     490.76 |     506.06 |  +3.12% |
| Latency 50th (ms)      |       0.65 |       0.64 |  -1.24% |
| Latency 75th (ms)      |       0.95 |       0.94 |  -1.05% |
| Latency 90th (ms)      |       3.27 |       5.97 | +82.57% |
| Latency 99th (ms)      |       0.00 |       0.00 |         |

**Comment by shirhatti** [comment]
Plaintext on `aspnet-citrine-lin`

| load                   | old        | new        |        |
| ---------------------- | ---------- | ---------- | ------ |
| CPU Usage (%)          |         19 |         19 |  0.00% |
| Cores usage (%)        |        531 |        531 |  0.00% |
| Working Set (MB)       |         37 |         37 |  0.00% |
| Private Memory (MB)    |        358 |        358 |  0.00% |
| Start Time (ms)        |          0 |          0 |        |
| First Request (ms)     |         70 |         72 | +2.86% |
| Requests/sec           |  2,476,644 |  2,475,423 | -0.05% |
| Requests               | 37,394,952 | 37,378,301 | -0.04% |
| Mean latency (ms)      |       1.15 |       1.15 |  0.00% |
| Max latency (ms)       |      49.90 |      53.67 | +7.56% |
| Bad responses          |          0 |          0 |        |
| Socket errors          |          0 |          0 |        |
| Read throughput (MB/s) |     311.77 |     311.62 | -0.05% |
| Latency 50th (ms)      |       0.99 |       0.96 | -3.03% |
| Latency 75th (ms)      |       1.46 |       1.41 | -3.42% |
| Latency 90th (ms)      |       2.07 |       1.91 | -7.73% |
| Latency 99th (ms)      |       0.00 |       0.00 |        |

**Comment by shirhatti** [comment]
@Tratcher The only new change I've made since we last spoke is to resolve the Propagator from DI instead of using the global static. Can you review that?

Also, the difference in the GC counters makes no sense to me 🤷. Can you help me make sense of that?

**Review by Tratcher — COMMENTED** [review]
The DI change seems fine

**Comment by Tratcher** [comment]
No idea on the GC data, but at least it's getting better and not worse 😁.

**Comment by shirhatti** [comment]
There's no way LOH is 672 bytes. Kestrel's slabbed memory pool is allocated in the LOH. @sebastienros any ideas? Repeating the run, I got 131800 for Max LOH size for both before/after.

As for reduction in Max Gen 1 and increase in Max Gen 2 size, I'm not sure what to make of it? Something is getting promoted sooner? I'll try collecting an allocation profile to make more sense of this

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (tarekgh) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:291]: > if (carrier is null) [](http://example.com/codeflow?start=19&length=21)

is this possible in yor case?

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (tarekgh) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:301]: >requestId is not null) [](http://example.com/codeflow?start=16&length=22)

does 

fieldValue = headers[fieldName];

return null or empty strings for non-found headers? is it possible it throws?

I am asking because we are checking only for null here.

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (tarekgh) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:305]: >if (traceState is not null) [](http://example.com/codeflow?start=12&length=27)

do we set the trace state if we didn't get request Id? I mean this block should be moved inside the previous block?

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (tarekgh) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:310]: >if (!string.IsNullOrEmpty(requestId)) [](http://example.com/codeflow?start=12&length=37)

can we move this block inside previous block checking on request Id?

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (tarekgh) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:291]: > if (carrier is null) [](http://example.com/codeflow?start=19&length=21)

may be you reverse this check like 

if (carrier is not null)
{
    var headers = (IHeaderDictionary)carrier;
    fieldValue = headers[fieldName];
    return;
}

so you initlaize fieldValue once instead of twice. this apply to baggage cal too.

Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:291]: Not that I'm aware. I did for the nullability warnings I got otherwise

**Comment by tarekgh** [comment]
>         private static HostingApplication CreateApplication(IHttpContextFactory httpContextFactory = null, bool useHttpContextAccessor = false,

May be a clever idea in a future PR to add a test consuming other propagator. e.g., tests create a custom propagator and consume it. You don't have to block this PR on that.

---
Refers to: src/Hosting/Hosting/test/HostingApplicationTests.cs:182 in abcc758. [](commit_id = abcc758e8e2255a96beaa1ffca03525a3f5c996e, deletion_comment = False)

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (tarekgh) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:291]: you can use `carrier!` to suppress that.

**Review by tarekgh — APPROVED** [review]
Added some commented. LGTM otherwise.

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:301]: Ahh yes. I didn't realize empty value HTTP headers are legal. Will update this

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:305]: What does the current propagator implementation do? Do you call the extract delegate for tracestate if you don't find requestid/traceparent?

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:310]: (y)

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:291]: I'll invert the if and add the null-forgiving operating

Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:291]: I meant remove the if

File: src/Hosting/Hosting/src/Microsoft.AspNetCore.Hosting.csproj
Comment (BrennanConroy) [inline src/Hosting/Hosting/src/Microsoft.AspNetCore.Hosting.csproj:11]: revert

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: Done

File: src/Hosting/Hosting/src/Microsoft.AspNetCore.Hosting.csproj
Comment (shirhatti) [reply src/Hosting/Hosting/src/Microsoft.AspNetCore.Hosting.csproj]: 😅

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (tarekgh) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: The propagator must try all fields regardless.

https://github.com/dotnet/runtime/blob/b412aac8757d7ac29ab66910e462649eeaecda18/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/LegacyPropagator.cs#L54

My comment here regarding asp.net compatibility with previous behavior. So, regardless of the propagator behavior we should keep asp.net old behavior.

Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: Done. Moved the check into the `!string.IsNullOrEmpty(requestId)` block

Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: @tarekgh Can you confirm I addressed your concern?

Comment (tarekgh) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: Yes, it looks good to me.

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (Tratcher) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:310]: You used to add these in reverse, does ExtractBaggage handle the reversal? If so then include that in the comment. Otherwise this comment about order sensitivity doesn't make sense.

**Review by Tratcher — APPROVED** [review]

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
Comment (shirhatti) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:310]: I’ll send a follow-up PR to fix the comment. I want to merge this when the PR checks are green in case AzDo has a bad day again

**Comment by davidfowl** [comment]
I'm assuming there's no extra allocations here.

**Pull request merged**
